Add response error message output to HTTP Status 401 Errors in FileDownloader

shoeffner opened the pull request:
Changelog: Fix: Add response error message output to HTTP Status 401 Errors in FileDownloader.
Docs: Omit


(was #15981, but I committed with the wrong mail address so the GPG signatures were broken)

The source download step attempts to load source_credentials.json when it is present. When following the recommended steps of passing credentials to the source_credentials.json via os.getenv('BACKUP_USER') etc., this can result in the json string "None" when no authentication is provided.
This effectively gets translated to the curl call when being read:

    curl -i -u None:None https://artifactory.example.org/artifactory/source-backup/

This causes an error with status code 401, which currently does not propagate the error message and is slightly misleading, as a source_credentials.json is provided but the content is wrong:

    ERROR: conanfile.py (example/1.0): Error in source() method, line 17
            get(self, self.url, sha256=sha256, strip_root=True)
            ConanException: The source backup server 'https://artifactory.example.org/artifactory/source-backup/' needs authentication: . Please provide 'source_credentials.json'

With this fix, the error message is propagated in the same way others in the same block, resulting in a more useful message, although the "please provide" is still misleading.

    ERROR: conanfile.py (example/1.0): Error in source() method, line 17
            get(self, self.url, sha256=sha256, strip_root=True)
            ConanException: The source backup server 'https://artifactory.example.org/artifactory/source-backup/' needs authentication: {
      "errors" : [ {
        "status" : 401,
        "message" : "Bad credentials"
      } ]
    }. Please provide 'source_credentials.json'


- [x] Refer to the issue that supports this Pull Request: Closes: #15982
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop2/.github/CONTRIBUTING.md).

(after creating the PR... sorry for not following the exact procedure of waiting for feedback etc.)

- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

shoeffner committed:
add response error message to 401 error in file downloader

The source download step attempts to load source_credentials.json when it is present.
When following the recommended steps of passing credentials to the source_credentials.json via
os.getenv('BACKUP_USER') etc., this can result in the json string "None" when
no authentication is provided.
This effectively gets translated to the curl call when being read:

    curl -i -u None:None https://artifactory.example.org/artifactory/source-backup/

This causes an error with status code 401, which currently does not propagate
the error message and is slightly misleading, as a source_credentials.json is
provided but the content is wrong:

    ERROR: conanfile.py (example/1.0): Error in source() method, line 17
            get(self, self.url, sha256=sha256, strip_root=True)
            ConanException: The source backup server 'https://artifactory.example.org/artifactory/source-backup/' needs authentication: . Please provide 'source_credentials.json'

With this fix, the error message is propagated in the same way others in the
same block, resulting in a more useful message, although the "please provide"
is still misleading.

    ERROR: conanfile.py (example/1.0): Error in source() method, line 17
            get(self, self.url, sha256=sha256, strip_root=True)
            ConanException: The source backup server 'https://artifactory.example.org/artifactory/source-backup/' needs authentication: {
  "errors" : [ {
    "status" : 401,
    "message" : "Bad credentials"
  } ]
}. Please provide 'source_credentials.json'

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=15983) <br/>All committers have signed the CLA.

memsharded reviewed (approved):
Would be great to capture this message in one test. I can contribute this next week if you cannot figure out how/where to do it.

Thanks for your contribution!

shoeffner committed:
add missing test dependencies PyJWT and pluginbase

shoeffner committed:
add tests for unauthorized downloads due to missing or bad credentials

Also change the TestFileServer to distinguish between the two cases
so that we can test them by replacing the bad credentials error
message "Not authorized" with "Bad credentials".

shoeffner commented:
I sorted the commits in the following order:
- I added two additional requirements missing from the dev_requirements which I needed to run the tests (PyJWT and pluginbase).
- I added (failing) tests and changed the response of the TestFileServer to contain "Bad credentials" if there are bad credentials to distinguish between missing and bad credentials (there was no distinction before)
- I re-added my changes to make the tests pass

I am also fighting with my git settings, hence the "unverified" commits in between... Now they should all be signed properly.

memsharded commented on conans/requirements_dev.txt:
```diff
@@ -4,3 +4,5 @@ parameterized>=0.6.3
 mock>=1.3.0, <1.4.0
 WebTest>=2.0.18, <2.1.0
 bottle
+PyJWT
```
Good, we used to install ``requirements_server.txt`` for this, but I agree that this shouldn't be necessary and installing the ``dev`` requirements should be enough to run the tests too.

memsharded commented on conans/test/unittests/tools/files/test_downloads.py:
```diff
@@ -77,6 +77,27 @@ def test_download_forbidden(self, _manual):
             download(conanfile, file_server.fake_url + "/forbidden", dest)
         assert "403 Forbidden" in str(exc.value)
 
+    def test_download_unauthorized_no_credentials(self, _manual):
```
Excellent. Adding a new test instead of modifying existing ones is generally better, nice one.

shoeffner replied:
Thank you!

memsharded reviewed (approved):
Many thanks for this contribution!
